File: workspace/python/itrack/frontend/src/types/entity.ts
export type EntityType = 'Home' | 'Office' | 'Custom';
export type MemberRole = 'admin' | 'member';
export type TransactionMode = 'shared' | 'private';

export interface EntityMember {
  user_id: string;
  username: string;
  role: MemberRole;
  joined_at: string;
}

export interface Entity {
  id: string;
  name: string;
  entity_type: EntityType;
  custom_type_name?: string;
  description?: string;
  members: EntityMember[];
  created_by: string;
  created_at: string;
}

export interface EntityCreate {
  name: string;
  entity_type: EntityType;
  custom_type_name?: string;
  description?: string;
}

export interface EntityUpdate {
  name?: string;
  entity_type?: EntityType;
  custom_type_name?: string;
  description?: string;
}

export interface EntityInvite {
  user_email: string;
  role: MemberRole;
}

export interface MemberBreakdown {
  username: string;
  income: number;
  expense: number;
  balance: number;
}

export interface EntitySummary {
  entity_id: string;
  entity_name: string;
  total_balance: number;
  total_income: number;
  total_expense: number;
  shared_balance: number;
  shared_income: number;
  shared_expense: number;
  transaction_count: number;
  shared_transaction_count: number;
  categories_breakdown: Record<string, number>;
  member_breakdown: Record<string, MemberBreakdown>;
}

File: workspace/python/itrack/backend/app/models/entity.py
from datetime import datetime, timezone
from typing import Optional, List, Literal
from pydantic import BaseModel, ConfigDict, Field
from app.models.base import PyObjectId


EntityType = Literal["Home", "Office", "Custom"]
MemberRole = Literal["admin", "member"]


class EntityMember(BaseModel):
    user_id: str
    role: MemberRole
    joined_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class EntityBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    entity_type: EntityType
    custom_type_name: Optional[str] = Field(None, max_length=50)
    description: Optional[str] = Field(None, max_length=500)


class EntityCreate(EntityBase):
    pass


class EntityUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    entity_type: Optional[EntityType] = None
    custom_type_name: Optional[str] = Field(None, max_length=50)
    description: Optional[str] = Field(None, max_length=500)


class EntityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    entity_type: EntityType
    custom_type_name: Optional[str] = None
    description: Optional[str] = None
    members: List[EntityMember]
    created_by: str
    created_at: datetime
    member_count: int


class EntityMemberResponse(BaseModel):
    user_id: str
    username: str
    email: str
    role: MemberRole
    joined_at: datetime


class EntityInviteRequest(BaseModel):
    user_email: str
    role: MemberRole = "member"


class EntitySummary(BaseModel):
    entity_id: str
    entity_name: str
    total_balance: float
    total_income: float
    total_expense: float
    total_assets: float = 0.0
    total_liabilities: float = 0.0
    net_worth: float = 0.0
    shared_balance: float
    shared_income: float
    shared_expense: float
    shared_assets: float = 0.0
    shared_liabilities: float = 0.0
    transaction_count: int
    shared_transaction_count: int
    categories_breakdown: dict[str, float]
    member_breakdown: dict[str, dict]
